v2/yarpcthrift/inbound.go

```diff
@@ -0,0 +1,154 @@
+	ctx, call := yarpc.NewInboundCall(ctx)
+
+	reqValue, responder, err := decodeRequest(call, req, reqBuf, t.Protocol, t.Enveloping)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	thriftRes, err := t.ThriftHandler(ctx, reqValue)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if resType := thriftRes.Body.EnvelopeType(); resType != wire.Reply {
+		return nil, nil, yarpcencoding.ResponseBodyEncodeError(
+			req, errUnexpectedEnvelopeType(resType))
+	}
+
+	value, err := thriftRes.Body.ToWire()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	res, resBuf := &yarpc.Response{}, &yarpc.Buffer{}
+
+	if thriftRes.IsApplicationError {
+		res.ApplicationError = true
+	}
+
+	call.WriteToResponse(res)
+
+	if err = responder.EncodeResponse(value, wire.Reply, resBuf); err != nil {
+		return nil, nil, yarpcencoding.ResponseBodyEncodeError(req, err)
+	}
+
+	return res, resBuf, nil
+}
+
+// decodeRequest is a utility shared by Unary handlers, to decode
+// the request, regardless of enveloping.
+func decodeRequest(
+	// call is an inboundCall populated from the transport request and context.
+	call *yarpc.InboundCall,
+	req *yarpc.Request,
+	reqBuf *yarpc.Buffer,
+	// proto is the encoding protocol (e.g., Binary) or an
+	// EnvelopeAgnosticProtocol (e.g., EnvelopeAgnosticBinary)
+	proto protocol.Protocol,
+	// enveloping indicates that requests must be enveloped, used only if the
+	// protocol is not envelope agnostic.
+	enveloping bool,
+) (
+	// the wire representation of the decoded request.
+	// decodeRequest does not surface the envelope.
+	wire.Value,
+	// how to encode the response, with the enveloping
+	// strategy corresponding to the request.
+	protocol.Responder,
+	error,
+) {
+	if err := yarpcencoding.ExpectEncodings(req, Encoding); err != nil {
+		return wire.Value{}, nil, err
+	}
+
+	if err := call.ReadFromRequest(req); err != nil {
+		// not reachable
+		return wire.Value{}, nil, err
+	}
+
+	// TODO(mhp): This will be unnecessary when `yarpc.Buffer` implements `io.ReaderAt`
+	reader := bytes.NewReader(reqBuf.Bytes())
+
+	// Discover or choose the appropriate envelope
+	if agnosticProto, ok := proto.(protocol.EnvelopeAgnosticProtocol); ok {
+		return agnosticProto.DecodeRequest(wire.Call, reader)
+	}
+	if enveloping {
+		return decodeEnvelopedRequest(req, proto, reader)
+	}
+	return decodeUnenvelopedRequest(proto, reader)
+}
+
+func decodeEnvelopedRequest(
+	req *yarpc.Request,
+	proto protocol.Protocol,
+	reader *bytes.Reader,
+) (wire.Value, protocol.Responder, error) {
+	var envelope wire.Envelope
+	envelope, err := proto.DecodeEnveloped(reader)
+	if err != nil {
+		return wire.Value{}, nil, err
+	}
+	if envelope.Type != wire.Call {
+		err := yarpcencoding.RequestBodyDecodeError(req, errUnexpectedEnvelopeType(envelope.Type))
+		return wire.Value{}, nil, err
+	}
+	reqValue := envelope.Value
+	responder := protocol.EnvelopeV1Responder{Name: envelope.Name, SeqID: envelope.SeqID}
+	return reqValue, responder, nil
```

Comment:
nit: Since `reqValue` isn't used elsewhere we could remove and update this to
```
return envelope.Value, responder, nil
```

Also fine with leaving as is if we prefer the "declare all of the return values above".